common/src/main/java/io/netty/util/internal/DefaultPriorityQueue.java

```diff
@@ -0,0 +1,291 @@
+            // "small", then grow by 50% when "large".
+            queue = Arrays.copyOf(queue, queue.length + ((queue.length < 64) ?
+                                                         (queue.length + 2) :
+                                                         (queue.length >>> 1)));
+        }
+
+        bubbleUp(size++, e);
+        return true;
+    }
+
+    @Override
+    public T poll() {
+        if (size == 0) {
+            return null;
+        }
+        T result = queue[0];
+        result.priorityQueueIndex(this, INDEX_NOT_IN_QUEUE);
+
+        T last = queue[--size];
+        queue[size] = null;
+        if (size != 0) { // Make sure we don't add the last element back.
+            bubbleDown(0, last);
+        }
+
+        return result;
+    }
+
+    @Override
+    public T peek() {
+        return (size == 0) ? null : queue[0];
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public boolean remove(Object o) {
+        final T node;
+        try {
+            node = (T) o;
+        } catch (ClassCastException e) {
+            return false;
+        }
+        return removeTyped(node);
+    }
+
+    @Override
+    public boolean removeTyped(T node) {
+        int i = node.priorityQueueIndex(this);
+        if (!contains(node, i)) {
+            return false;
+        }
+
+        node.priorityQueueIndex(this, INDEX_NOT_IN_QUEUE);
+        if (--size == 0 || size == i) {
+            // If there are no node left, or this is the last node in the array just remove and return.
+            queue[i] = null;
+            return true;
+        }
+
+        // Move the last element where node currently lives in the array.
+        T moved = queue[i] = queue[size];
+        queue[size] = null;
+        // priorityQueueIndex will be updated below in bubbleUp or bubbleDown
+
+        // Make sure the moved node still preserves the min-heap properties.
+        if (comparator.compare(node, moved) < 0) {
+            bubbleDown(i, moved);
+        } else {
+            bubbleUp(i, moved);
+        }
+        return true;
+    }
+
+    @Override
+    public void priorityChanged(T node) {
+        int i = node.priorityQueueIndex(this);
+        if (!contains(node, i)) {
+            return;
+        }
+
+        // Preserve the min-heap property by comparing the new priority with parents/children in the heap.
+        if (i == 0) {
+            bubbleDown(i, node);
+        } else {
+            // Get the parent to see if min-heap properties are violated.
+            int iParent = (i - 1) >>> 1;
+            T parent = queue[iParent];
+            if (comparator.compare(node, parent) < 0) {
+                bubbleUp(i, node);
+            } else {
+                bubbleDown(i, node);
+            }
+        }
+    }
+
+    @Override
+    public Object[] toArray() {
+        return Arrays.copyOf(queue, size);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <X> X[] toArray(X[] a) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)